kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
-        } else if test_flag == 0x55 {
-            info!("ps2 controller test pass!!!")
-        }
+    write_command(TestController);
+    match read_controller_test_result()? {
+        ControllerTestResult::Passed => debug!("passed PS/2 controller test"),
+        ControllerTestResult::Failed => Err("failed PS/2 controller test")?,
     }
 
-    // test the first ps2 data port
-    {
-        ps2_write_command(0xAB);
-        let test_flag = ps2_read_data();
-        match test_flag {
-            0x00 => info!("first ps2 port test pass!!!"),
-            0x01 => warn!("first ps2 port clock line stuck low"),
-            0x02 => warn!("first ps2 port clock line stuck high"),
-            0x03 => warn!("first ps2 port data line stuck low"),
-            0x04 => warn!("first ps2 port data line stuck high"),
-            _ => {}
-        }
+    // test the ps2 data port
+    match port {
+        PS2Port::One => write_command(TestPort1),
+        PS2Port::Two => write_command(TestPort2),
+    }
+    use PortTestResult::*;
+    match read_port_test_result()? {
+        Passed => debug!("passed PS/2 port {} test", port as u8 + 1),
+        ClockLineStuckLow => Err("failed PS/2 port test, clock line stuck low")?,
+        ClockLineStuckHigh => Err("failed PS/2 port test, clock line stuck high")?,
+        DataLineStuckLow => Err("failed PS/2 port test, data line stuck low")?,
+        DataLineStuckHigh => Err("failed PS/2 port test, clock line stuck high")?,
     }
 
     // enable PS2 interrupt and see the new config
-    {
-        let config = ps2_read_config();
-        if config & 0x01 == 0x01 {
-            info!("first ps2 port interrupt enabled")
-        } else {
-            warn!("first ps2 port interrupt disabled")
-        }
-        if config & 0x10 != 0x10 {
-            info!("first ps2 data port enabled")
-        } else {
-            warn!("first ps2 data port disabled")
-        }
+    let config = read_config();
+    let port_interrupt_enabled = match port {
+        PS2Port::One => config.port1_interrupt_enabled(),
+        PS2Port::Two => config.port2_interrupt_enabled(),
+    };
+    let clock_enabled = match port {
+        PS2Port::One => !config.port1_clock_disabled(),
+        PS2Port::Two => !config.port2_clock_disabled(),
+    };
+
+    if port_interrupt_enabled {
+        debug!("ps2 port {} interrupt enabled", port as u8 + 1)
+    } else {
+        Err("PS/2 port test config's interrupt disabled")?
     }
+    if clock_enabled {
+        debug!("ps2 port {} clock enabled", port as u8 + 1)
+    } else {
+        Err("PS/2 port test config's clock disabled")?
+    }
+    Ok(())
 }
 
-/// initialize the second ps2 data port
-pub fn init_ps2_port2() {
-    //disable PS2 ports first
-    {
-        ps2_write_command(0xA7);
-        ps2_write_command(0xAD);
-    }
+#[derive(TryFromPrimitive)]
+#[repr(u8)]
+enum ControllerTestResult {
+    Passed = 0x55,
+    //NOTE: can just be removed
```

Comment:
why can this be removed? If so, should we just remove it now?

Reply:
probably missed that one, here the spec says "Failed is 0xFC", but it doesn't matter because
`read_data().try_into().map_err(|_| "failed to read controller test result")`
rejects anything besides "Passed" anyways.

Reply:
got it, makes sense. Feel free to remove it.